src/D2L.CodeStyle.Analyzers/Helpers/WhitelistHelper.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+
+namespace D2L.CodeStyle.Analyzers.Helpers {
+
+	internal static class WhitelistHelper {
+
+		public static WhitelistHelper<INamedTypeSymbol> TypeLevel(
+			string whitelistFileName
+		) => new WhitelistHelper<INamedTypeSymbol>(
+			whitelistFileName: whitelistFileName,
+			entryFormatter: s => $"{ s.ToString() }, { s.ContainingAssembly.ToDisplayString( SymbolDisplayFormat.MinimallyQualifiedFormat ) }"
+		);
+
+	}
+
+	internal sealed class WhitelistHelper<T> {
+
+		public WhitelistHelper(
+			string whitelistFileName,
+			Func<T, string> entryFormatter
+		) {
+			WhitelistFileName = whitelistFileName;
+			EntryFormatter = entryFormatter;
+		}
+
+		public string WhitelistFileName { get; }
+		public Func<T, string> EntryFormatter { get; }
+
+		public Func<T, bool> LoadWhitelist(
+			AnalyzerOptions options
+		) {
+			ImmutableHashSet<string> whitelist = LoadWhitelist( options.AdditionalFiles );
+			return IsInWhitelist;
+
+			bool IsInWhitelist( T candidate ) {
+				string entry = EntryFormatter( candidate );
+				bool whitelisted = whitelist.Contains( entry );
+				return whitelisted;
+			}
+		}
+
+		public delegate void ReportDiagnostic( Diagnostic diagnostic );
+		public void EntryIsUnnecessary( T entry, Location location, ReportDiagnostic report ) {
+			report( Diagnostic.Create(
+				descriptor: Diagnostics.UnnecessaryWhitelistEntry,
+				location: location,
+				EntryFormatter( entry ),
+				WhitelistFileName
+			) );
+		}
+
+		private ImmutableHashSet<string> LoadWhitelist(

```

Comment:
Does this need to be `string` based?

Reply:
I assume you mean trying to load them into type infos? I don't have an intuitive sense of how expensive that would be to do each project.

Reply:
At least just the symbol part, but ya dno.

Reply:
I don't think whether the whitelist contains uses the formatted string or loaded symbols is a blocker, but avoiding strings as much as possible is cool. Definitely think it's something to continue looking at afterward.